### ISSUE TITLE
Supporting variadic length of categoricals

### ISSUE THREAD

**Issue by zhaodong-wang** [jdb78/pytorch-forecasting#840]
Hi, I am wondering if the model can support categoricals being a list of strings with varying lengths. I saw there is already support for combining multiple categoricals into one using embedding bags, but this is still a fixed length. Is this something that can be supported by some tricks or this is something that needs code changes in the model?

**Comment by ozanozyegen** [comment]
I don't think it is supported based on the original paper. None of the datasets that were used in the original paper had categoricals with varying lengths.

That said, it should be possible to apply a trick to support them. Have you tried a simple solution like filling the missing time indexes with a 'missing' category tag?

**Comment by zhaodong-wang** [comment]
Thanks for your suggestion @ozanozyegen!
Not sure if I understand your solution correctly, do you mean that I pad all these categoricals with varying lengths and make their lengths all the same?
That might be a bit memory inefficient since the lengths of those categoricals range from 10 to 10000, so there will be a lot of padding values.

**Comment by ozanozyegen** [comment]
Yes, pad the missing time_idxs with a special tag. The recurrent architectures in TFT already expects some kind of input for time varying features at each timestep. So, I wouldn't worry about the memory inefficiency unless you have 100s of gbs of data.
If all the features are missing at some time indexes, you can use `allow_missing_timesteps` parameter in [Time Series Dataset](https://pytorch-forecasting.readthedocs.io/en/stable/data.html).

**Comment by zhaodong-wang** [comment]
Got it, @ozanozyegen. I will give it a try. But I am still concerned about the performance if I just simply throw in a dateframe with 10000+ columns, where 10k columns are categorical.

**Comment by ozanozyegen** [comment]
If we are talking about the same thing, then you should have 10k rows where some category columns are filled with the special tags. I don't see why you need to throw 10k columns to the dataset. Can you give an example to make things more clear?

**Comment by zhaodong-wang** [comment]
@ozanozyegen  Sure, here is an example of the dataset:

| var_categorical            | categorical | continuous | time_idx |
|----------------------------|-------------|------------|----------|
| ["v1", "v2", ... "v10"]    | "a1"        | 1.0        | 0        |
| ["v1", "v2", ... "v10000"] | "a2"        | 2.0        | 1        |
| ["v1", "v2"]         | "a3"        | 3.0        | 2        |


The column "var_categorical" is the variadic categorical time-varying feature I am going to use for prediction. This feature contains bags of words with varying lengths (ranging from 2-10k). My understanding for the padding that you mentioned is to transform the table to the following:

| cat_1 | cat_2 | cat_3 | ... | cat_10000 | categorical | continuous | time_idx |
|-------|-------|-------|-----|-----------|-------------|------------|----------|
| "v1"  | "v2"  | "v3"  | ... | "X"       | "a1"        | 1.0        | 0        |
| "v1"  | "v2"  | "v3"  | ... | "v10000"  | "a2"        | 2.0        | 1        |
| "v1"  | "v2"  | "X"   | ... | "X"       | "a3"        | 3.0        | 2        |

where "X" is the special tag. Then `cat_1` to `cat_10000` can be combined into one categorical group using the provided API, and thus we have a 10k column table.

**Comment by ozanozyegen** [comment]
Thanks for the example. So, you have many binary categoricals at each time step. Like a sentence at each time step. This is an uncommon situation.
Ideally, you want to use a single embedding layer and combine the embeddings of the categories, similar to the [RETAIN](https://arxiv.org/abs/1608.05745) architecture. 

I would copy the model I want to use and pass all the `var_categoricals` through the same embedding layer and pass the sum of the embeddings to the rnn layer.

**Comment by zhaodongw** [comment]
Thanks @ozanozyegen for your suggestions! I will try to copy the model and combine it with a customized embedding layer for this use case then.
